File: backend/src/ai/providers/openai.ts
import type { EmbeddingProvider } from "../types";
import { createLogger } from "../../utils/logger";

const log = createLogger("ai-openai");

interface OpenAIEmbeddingsResponse {
  data: Array<{ embedding: number[]; index: number }>;
}

export class OpenAIEmbeddingProvider implements EmbeddingProvider {
  readonly name = "openai";
  readonly model: string;
  readonly dimensions: number;
  readonly cacheNamespace: string;

  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly maxAttempts: number;
  private readonly retryBaseDelayMs: number;
  private readonly batchSize: number;

  constructor(options: {
    apiKey: string;
    model: string;
    dimensions: number;
    baseUrl?: string;
    maxAttempts: number;
    retryBaseDelayMs: number;
    batchSize: number;
  }) {
    this.apiKey = options.apiKey;
    this.model = options.model;
    this.dimensions = options.dimensions;
    this.cacheNamespace = `${this.name}:${this.model}:${this.dimensions}`;
    this.baseUrl = options.baseUrl ?? "https://api.openai.com/v1";
    this.maxAttempts = Math.max(1, options.maxAttempts);
    this.retryBaseDelayMs = Math.max(50, options.retryBaseDelayMs);
    this.batchSize = Math.max(1, options.batchSize);
  }

  async embedOne(text: string): Promise<number[]> {
    const [embedding] = await this.embedBatch([text]);
    return embedding ?? [];
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];

    const allEmbeddings: number[][] = [];
    for (let start = 0; start < texts.length; start += this.batchSize) {
      const batch = texts.slice(start, start + this.batchSize);
      const embeddings = await this.embedBatchWithRetry(batch);
      allEmbeddings.push(...embeddings);
    }

    return allEmbeddings;
  }

  private async embedBatchWithRetry(texts: string[]): Promise<number[][]> {
    let attempt = 0;
    let lastError: Error | null = null;

    while (attempt < this.maxAttempts) {
      attempt += 1;
      try {
        return await this.embedBatchOnce(texts);
      } catch (error) {
        const err = error instanceof Error ? error : new Error("Embedding request failed");
        lastError = err;
        if (attempt >= this.maxAttempts) {
          break;
        }

        const status = (err as Error & { status?: number }).status;
        const retriable = typeof status === "number"
          ? status === 429 || status >= 500
          : true;
        if (!retriable) {
          break;
        }

        const backoffMs = this.retryBaseDelayMs * 2 ** (attempt - 1);
        log.warn({ attempt, maxAttempts: this.maxAttempts, backoffMs, err }, "embedding batch retrying");
        await new Promise((resolve) => setTimeout(resolve, backoffMs));
      }
    }

    throw lastError ?? new Error("Embedding request failed");
  }

  private async embedBatchOnce(texts: string[]): Promise<number[][]> {

    const response = await fetch(`${this.baseUrl}/embeddings`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: texts,
        dimensions: this.dimensions,
      }),
    });

    if (!response.ok) {
      const body = await response.text();
      const error = new Error(`Embedding API failed (${response.status}): ${body}`) as Error & { status?: number };
      error.status = response.status;
      throw error;
    }

    const payload = (await response.json()) as OpenAIEmbeddingsResponse;
    return payload.data.sort((a, b) => a.index - b.index).map((item) => item.embedding);
  }
}

File: backend/app/ai/providers/openai.py
from __future__ import annotations

import asyncio
import logging

import httpx

log = logging.getLogger("ai-openai")


class OpenAIEmbeddingProvider:
    name = "openai"

    def __init__(
        self,
        *,
        api_key: str,
        model: str,
        dimensions: int,
        base_url: str = "https://api.openai.com/v1",
        max_attempts: int = 4,
        retry_base_delay_ms: int = 300,
        batch_size: int = 64,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.dimensions = dimensions
        self.cache_namespace = f"{self.name}:{model}:{dimensions}"
        self.base_url = base_url.rstrip("/")
        self.max_attempts = max(1, max_attempts)
        self.retry_base_delay_ms = max(50, retry_base_delay_ms)
        self.batch_size = max(1, batch_size)

    async def embed_one(self, text: str) -> list[float]:
        results = await self.embed_batch([text])
        return results[0] if results else []

    async def embed_batch(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        all_embeddings: list[list[float]] = []
        for start in range(0, len(texts), self.batch_size):
            batch = texts[start : start + self.batch_size]
            embeddings = await self._embed_batch_with_retry(batch)
            all_embeddings.extend(embeddings)
        return all_embeddings

    async def _embed_batch_with_retry(self, texts: list[str]) -> list[list[float]]:
        last_error: Exception | None = None
        for attempt in range(1, self.max_attempts + 1):
            try:
                return await self._embed_batch_once(texts)
            except Exception as exc:
                last_error = exc
                if attempt >= self.max_attempts:
                    break
                status = getattr(exc, "status_code", None)
                retriable = status is None or status == 429 or status >= 500
                if not retriable:
                    break
                delay_s = self.retry_base_delay_ms * (2 ** (attempt - 1)) / 1000
                log.warning(
                    "embedding batch retrying attempt=%d max=%d delay=%.2fs err=%s",
                    attempt,
                    self.max_attempts,
                    delay_s,
                    exc,
                )
                await asyncio.sleep(delay_s)
        raise last_error or RuntimeError("Embedding request failed")

    async def _embed_batch_once(self, texts: list[str]) -> list[list[float]]:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/embeddings",
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {self.api_key}",
                },
                json={"model": self.model, "input": texts, "dimensions": self.dimensions},
                timeout=60.0,
            )
        if response.status_code != 200:
            err = RuntimeError(
                f"Embedding API failed ({response.status_code}): {response.text}"
            )
            err.status_code = response.status_code  # type: ignore[attr-defined]
            raise err
        payload = response.json()
        items = sorted(payload["data"], key=lambda x: x["index"])
        return [item["embedding"] for item in items]
